feat(crashdump): Add crash dump functionality for fatal errors

slurmlord opened the pull request:
This change introduces the option to generate crash dumps, aka. mini dumps, on fatal errors.
The main minidump functionality is done by explicitly loading the `dbghelp.dll` from the system directory, as the dbghelp.dll that is bundled with the game is an older version that does not include this functionality. There is an option to create small dumps or extended info dumps, currently both are created.


### Small dumps
These mostly contain stacks for the process threads and some stack variables, or to create dumps with extended info. The use case for these is to quickly determine where a crash occured, the type of crash, if it was already fixed etc. In addition, if the memory allocation structures are corrupted enough, an extended info dump might not succeed while the small dump should. The size of these dumps are typically on the order of 250kB.

### Extended info
These contain global values, along with the memory regions allocated via the memory pool factory and the dynamic memory allocator. This makes all in-game objects available to the person debugging the crash dump, so for example `dt generalszh!TheWritableGlobalData` in WinDbg will show the state at the time the dump was created.

An alternative option could be to not traverse the memory structures "manually" to get to the allocations and instead just specify the `MiniDumpWithFullMemory` flag to `MiniDumpWriteDump`, but that increases the file size considerably.

As an example, dump of the generalszh process in the main menu with the shell map in the background yields a ~140MB dump when traversing and ~420MB with `MiniDumpWithFullMemory`. Beyond that, the ~140MB file compresses to ~20MB with 7Z, so should be relatively easily transferable.

### Storage Location
Crash dumps are stored in a new folder called 'CrashDumps' under the userDir ("Documents\Command and Conquer Generals Zero Hour Data\"), and on startup it will create this directory if it doesn't exist and delete any older dumps so only the 10 newest small and 2 newest extended info dumps are left. This is to preserve disk space, as the extended info files can be several hundred MB.

### Integration points
For VS2022 builds, unhandled exceptions end up in the `UnhandledExceptionFilter` in WinMain, which then get a reference to the actual exception that occurred and includes that in the dump.
For VC6 builds, unhandled exceptions are caught in the `catch(...)` blocks of `GameEngine::execute` which then calls RELEASE_CRASH. As there is no exception data available in this case to populate _EXCEPTION_POINTERS from, an intentional exception is triggered to get the trace of the current thread. This makes the stack traces for VC6 a bit more cryptic than VS2022 builds as the C++ exception handling gets included in the trace.

### Limitations

In the longer run we'll probably want to replace this code with a more mature solution, like CrashPad, but that currently depends on a newer compiler than VC6.
As the code is intended to be temporary, it's kept behind a new CMake feature so it can be easily removed. There are also some other decisions made with this in mind:

* Minidump is created in-process. Ideally, the dump should be performed by a process external to the crashing/failing process, but to avoid having to ship an extra binary, in-process was chosen instead. It's being performed in a separate thread to hopefully have a clean stack to work with. 
* Depends on RTS_BUILD_OPTION_VC6_FULL_DEBUG for VC6 builds. The PDBs generated with the default VC6 compile options are lacking in information, making the mini dumps less useful. The option RTS_BUILD_OPTION_VC6_FULL_DEBUG should be enabled for VC6 builds to ensure maximum usability. VS2022 builds produce better PDBs and require no extra options.
* Directory management code is contained within the MiniDumper class, not re-usable by other components.
* Many Win32-specific types and functions are used directly without regards for portability.
* As the MiniDump feature is not available for VC6, a lot of headers have been borrowed from minidumpapiset.h and included in the MiniDumper_compat.h file.
* Globals are used for storing the current exception info.
* Only enabled for the games, tools are currently not included.

xezon commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -0,0 +1,725 @@
+
+	// Use placement new on the process heap so TheMiniDumper is placed outside the MemoryPoolFactory managed area.
+	// If the crash is due to corrupted MemoryPoolFactory structures, try to mitigate the chances of MiniDumper memory also being corrupted
+	TheMiniDumper = new (::HeapAlloc(::GetProcessHeap(), HEAP_GENERATE_EXCEPTIONS, sizeof(MiniDumper))) MiniDumper;
```
After #1066 is merged, you can also use `sysAllocateDoNotZero` / `sysFree` for it.

xezon commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -0,0 +1,725 @@
+		// Only include data segments for the game and ntdll modules to keep dump size low
+		if (output.ModuleWriteFlags & ModuleWriteDataSeg)
+		{
+			if (::StrCmpIW(input.Module.FullPath, m_executablePath) != 0 && !::StrStrIW(input.Module.FullPath, L"ntdll.dll"))
```
Inconsistent comparison style

xezon commented on Core/Libraries/Source/WWVegas/WWLib/DbgHelpLoader.h:
```diff
@@ -160,6 +185,9 @@ class DbgHelpLoader
 		PTRANSLATE_ADDRESS_ROUTINE TranslateAddress);
 
+#ifdef RTS_ENABLE_CRASHDUMP
+	MiniDumpWriteDump_t m_miniDumpWriteDump;
```
Maybe add it to the end of the functions? It does make sense to have `m_symInitialize` first.

slurmlord replied:
Yep, I went with the alphabetical ordering initially but as it's an optional feature it reads better in a separate area.

slurmlord commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -148,36 +145,12 @@ void MiniDumper::TriggerMiniDumpForException(struct _EXCEPTION_POINTERS* e_info,
-	// Find the full path to the dbghelp.dll file in the system32 dir
-	::GetSystemDirectory(m_sysDbgHelpPath, MAX_PATH);
-	strlcat(m_sysDbgHelpPath, "\\dbghelp.dll", MAX_PATH);
+	bool success = DbgHelpLoader::load();
```
Looks good, I'll wait for #1804 to be merged before pushing a new revision as the `DbgHelpLoader::freeResources` method needs to be updated with the `m_miniDumpWriteDump` reference.

slurmlord replied:
I'm actively working on this, even though no new revision has been pushed yet.
For some reason dumps with the flag `MiniDumpWithDataSegs` set end up with corrupted stacks. I'm starting to suspect the OS (dbghelp/dbgeng/dbgcore), but not certain yet.
In any case it makes validation hard, delaying things a bit.

slurmlord replied:
Figured out the broken stack issue; specifying the dump flag `MiniDumpWithDataSegs` without also specifying `MiniDumpWithPrivateReadWriteMemory` results in a broken stack.
Can't remember having used `MiniDumpWithPrivateReadWriteMemory` before, but I might be misremembering.

slurmlord replied:
It's resolved - the solution was to add the `MiniDumpWithPrivateReadWriteMemory` flag to the MiniDumpWriteDump call, which was fixed in the last iteration. Based on my testing I would say this is ready.

slurmlord commented on Core/Libraries/Source/WWVegas/WWLib/DbgHelpLoader.h:
```diff
@@ -177,13 +202,15 @@ class DbgHelpLoader
 	Processes m_initializedProcesses;
 	HMODULE m_dllModule;
 	int m_referenceCount;
-	CriticalSectionClass m_criticalSection;
```
Took the liberty of removing this as it was unused.

xezon reviewed (commented):
I think the cleanup in MiniDumper class needs another look. Make it simple and robust.

slurmlord commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -0,0 +1,707 @@
+	}
+
+	::SetEvent(m_quitting);
+	DWORD waitRet = ::WaitForSingleObject(m_dumpThread, 3000);
```
Yep - looks better now. The expectation now is that if initMiniDumper has been called then shutdownMiniDumper should also be called, regardless of the result from init.

xezon commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -0,0 +1,699 @@
+#if RTS_GENERALS
+	Char product = 'G';
+#elif RTS_ZEROHOUR
+	Char product = 'Z';
```
Can make const

xezon reviewed (dismissed):
Code looks good to me. A few minor comments that could be looked into.

Caball009 commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -0,0 +1,704 @@
+		sysTime.wDay, sysTime.wHour, sysTime.wMinute, sysTime.wSecond,
+		GitShortSHA1, currentProcessId);
+
+	HANDLE dumpFile = CreateFile(m_dumpFile, GENERIC_READ | GENERIC_WRITE, 0, NULL, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, NULL);
```
::CreateFile

Caball009 commented on Core/GameEngine/Include/Common/MiniDumper.h:
```diff
@@ -0,0 +1,113 @@
+	MiniDumper();
+	Bool IsInitialized() const;
+	void TriggerMiniDump(DumpType dumpType);
+	void TriggerMiniDumpForException(struct _EXCEPTION_POINTERS* e_info, DumpType dumpType);
```
Can we get rid of the `struct` keyword anywhere that's not a definition or declaration of the struct itself?

xezon commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -0,0 +1,732 @@
+Bool MiniDumper::ShouldWriteDataSegsForModule(const PWCHAR module) const
+{
+	// Only include data segments for the game, ntdll and kernel32 modules to keep dump size low
+	constexpr const static WideChar* wanted_modules[] = { L"ntdll.dll", L"kernel32.dll"};
```
Is this order legal? Remove "static".

xezon replied:
Ok fair point about VC6. Make it `static constexpr const`.

xezon commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -0,0 +1,732 @@
+	}
+}
+
+bool MiniDumper::EndsWithCaseInsensitive(const WideChar* str, const WideChar* suffix)
```
String helper functions would be a better fit for stringex.h in WWLib.

Perhaps add new general purpose endswith functions like AsciiString has them?

AsciiString::startsWith
AsciiString::startsWithNoCase
AsciiString::endsWith
AsciiString::endsWithNoCase

Implement them standalone for both const char* and const wchar_t*

Then also use them in AsciiString and UnicodeString. I can help you with that if needed. Better make as a separate change (another one haha).

xezon commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -0,0 +1,710 @@
+	case ThreadCallback:
+		retVal = TRUE;
+		break;
+	case ThreadExCallback:
```
Can merge ThreadCallback, ThreadCallback, ThreadExCallback cases.

xezon commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -0,0 +1,710 @@
+// This is where the memory regions and things are being filtered
+BOOL MiniDumper::CallbackInternal(const MINIDUMP_CALLBACK_INPUT& input, MINIDUMP_CALLBACK_OUTPUT& output)
+{
+	BOOL retVal = TRUE;
```
What does the `retVal` signal to the caller? Perhaps give it a name that describes it, because it is not obvious what return TRUE and FALSE will do now.

slurmlord commented:
Refactored a bit, introducing an enum for the memory range phases and modified the output variable of the callback routine so that our callback can always return TRUE.

xezon commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -587,66 +580,61 @@ BOOL MiniDumper::DumpMemoryObjects(ULONG64& memoryBase, ULONG& memorySize)
 		}
 		break;
 	}
 	default:
```
Maybe put

```cpp
case COMPLETED:
  ...
  break;
default:
  DEBUG_CRASH("Invalid state");
  break;
```

xezon commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -0,0 +1,698 @@
+			input.ReadMemoryFailure.Offset, input.ReadMemoryFailure.Bytes, input.ReadMemoryFailure.FailureStatus));
+		break;
+	case CancelCallback:
+		output.Cancel = FALSE;
```
What does this do?

xezon commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -0,0 +1,698 @@
+		if (m_dumpObjectsSubState == 0)
+		{
+			m_rangeIter = TheMemoryPoolFactory->cbegin();
+			++m_dumpObjectsSubState;
```
Perhaps can get rid of `m_dumpObjectsSubState` by simply initializing `m_rangeIter = TheMemoryPoolFactory->cbegin()` before this callback?

xezon commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -0,0 +1,698 @@
+		if (m_dumpObjectsSubState < rawBlocksInDma)
+		{
+			// Dump this block
+			m_currentAllocator->fillAllocationRangeForRawBlockN(m_dumpObjectsSubState, rawBlockRange);
```
`fillAllocationRangeForRawBlockN` also has a loop inside.

Can we perhaps cache the MemoryPoolSingleBlock* pointer and just keep going to next one until it is NULL, without using `getRawBlockCount`, `fillAllocationRangeForRawBlockN`?

xezon reviewed (commented):
To me it looks like we can get rid of the concept of `m_dumpObjectsSubState` by caching the actual pointers to memory structures and keep iterating until they reach NULL and then go to the next until all lists reached the end.

Performance wise it is probably not the biggest of deals but it should make the logic cleaner.

slurmlord commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -0,0 +1,698 @@
+		{
+			// DumpMemoryObjects will set outputMemorySize to 0 once it's completed, signalling the end of memory callbacks
+			DumpMemoryObjects(output.MemoryBase, output.MemorySize);
+		} while ((output.MemoryBase == 0 || output.MemorySize == 0) && m_dumpObjectsState != COMPLETED);
```
It's valid programming wise, but not as a callback-value. Returning a region with size 0 means "done with memory regions", so no more `MemoryCallback`s will be sent after that.
That's why, when transitioning between the phases in `DumpMemoryObjects` we need to call it again if `output.MemorySize` was 0 so we don't stop getting `MemoryCallback`s. The check for `COMPLETED` is so we know when we should actually return the 0 sized region.
Checking `MemoryBase` is not needed, I'll remove that.

slurmlord commented:
Got rid of the `m_dumpObjectsSubState` and cached the various objects that's being iterated. This also reduced the changes required in `GameMemory`. Looks better!

xezon commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -495,7 +496,7 @@ BOOL MiniDumper::CallbackInternal(const MINIDUMP_CALLBACK_INPUT& input, MINIDUMP
 			// DumpMemoryObjects will set outputMemorySize to 0 once it's completed, signalling the end of memory callbacks
 			DumpMemoryObjects(output.MemoryBase, output.MemorySize);
-		} while ((output.MemoryBase == 0 || output.MemorySize == 0) && m_dumpObjectsState != COMPLETED);
+		} while (output.MemorySize == 0 && m_dumpObjectsState != COMPLETED);
```
Shouldn't this be `!= 0` then?

slurmlord replied:
I moved this condition into `DumpMemoryObjects` which also made it not depend on `m_dumpObjectsState` outside of that method.
Hopefully the do-while along with the comment in `DumpMemoryObjects` makes sense.

xezon commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -495,7 +496,7 @@ BOOL MiniDumper::CallbackInternal(const MINIDUMP_CALLBACK_INPUT& input, MINIDUMP
 		{
 			// DumpMemoryObjects will set outputMemorySize to 0 once it's completed, signalling the end of memory callbacks
```
"memory size"

xezon commented on Core/GameEngine/Include/Common/MiniDumper.h:
```diff
@@ -0,0 +1,124 @@
+	DUMP_TYPE_FULL = 'F',
+};
+
+enum MiniDumperExitCode CPP_11(: Int)
```
Can this enum be private in MiniDumper class?

xezon commented on Core/GameEngine/Include/Common/MiniDumper.h:
```diff
@@ -0,0 +1,124 @@
+	DUMPER_EXIT_FORCED_TERMINATE = 0x158B1154,
+};
+
+enum DumpObjectsState CPP_11(: Int)
```
Can this enum be private in MiniDumper class?

xezon commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -0,0 +1,667 @@
+		if (m_currentSingleBlock == NULL)
+		{
+			// Iterated to a new allocator, start iterating over its blocks
+			m_currentSingleBlock = m_currentAllocator->getFirstRawBlock();
```
Remove this.

xezon reviewed (commented):
This looks much cleaner indeed.

xezon commented on Core/GameEngine/Include/Common/MiniDumper.h:
```diff
@@ -31,25 +31,25 @@ enum DumpType CPP_11(: Char)
+	{
+		DumpObjectsState_Begin,
+		DumpObjectsState_Memory_Pools,
+		DumpObjectsState_Memory_Pool_Allocations,
```
Minor: Prefer `DumpObjectsState_MemoryPoolAllocations`, where underscore only separates enum type name from value name.

xezon commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -520,90 +518,89 @@ BOOL MiniDumper::CallbackInternal(const MINIDUMP_CALLBACK_INPUT& input, MINIDUMP
+			return;
+		default:
+			DEBUG_CRASH(("Invalid object state"));
+			break;
```
Better make this exit too to avoid infinite loop if this error ever happened.

xezon commented on Core/GameEngine/Include/Common/GameMemory.h:
```diff
@@ -474,13 +488,81 @@ class DynamicMemoryAllocator
+	};
+
+	reference operator*() { UpdateRange(); return m_range; }
+	pointer operator->() { UpdateRange(); return &m_range; }
```
It looks like `UpdateRange()` should happen elsewhere, for example in the increment. Because dereferencing should not update anything. It should just give reference.

xezon commented on Core/GameEngine/Source/Common/System/GameMemory.cpp:
```diff
@@ -3232,6 +3267,49 @@ void MemoryPoolFactory::debugMemoryReport(Int flags, Int startCheckpoint, Int en
+	m_currentBlobInPool->fillAllocatedRange(m_range);
+}
+
+void AllocationRangeIterator::MoveToNextBlob()
```
Some function names here are not camelCase

xezon commented on Core/GameEngine/Source/Common/System/GameMemory.cpp:
```diff
@@ -3232,6 +3268,114 @@ void MemoryPoolFactory::debugMemoryReport(Int flags, Int startCheckpoint, Int en
+}
+
+// Prefix increment
+AllocationRangeIterator& AllocationRangeIterator::operator++() { moveToNextBlob(); updateRange(); return *this; }
```
style:

```cpp
{
  moveToNextBlob();
  updateRange();
  return *this;
}
```

xezon commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -520,90 +518,89 @@ BOOL MiniDumper::CallbackInternal(const MINIDUMP_CALLBACK_INPUT& input, MINIDUMP
-		}
+			// Iterate through all the direct allocations ("raw blocks") done by DMAs, as these are done outside of the
+			// memory pool factory allocations dumped in the previous phase.
+			if (m_currentAllocator == NULL || m_currentSingleBlock == NULL)
```
You can set `m_currentAllocator` back to NULL if `m_currentAllocator->getFirstRawBlock()` returned NULL.

xezon commented on Core/GameEngine/Include/Common/GameMemory.h:
```diff
@@ -474,13 +488,52 @@ class DynamicMemoryAllocator
+
+	void updateRange();
+	void moveToNextBlob();
+	const MemoryPoolFactory* m_factory;
```
This member looks not necessary

xezon commented on Core/GameEngine/Source/Common/System/MiniDumper.cpp:
```diff
@@ -0,0 +1,682 @@
+	}
+}
+
+void MiniDumper::MoveToNextSingleBlock()
```
`MoveToNextAllocatorWithRawBlocks` and `MoveToNextSingleBlock` can likely be one function like how you did it in the Memory Pool Iterator.

slurmlord commented:
I'm considering to reduce the scope of this PR a bit to make it easier to review etc.
The proposal is to remove the `GameMemory` dump mode, and only have `Small` and `Full` in this PR. GameMemory can be added in a subsequent PR if we see that the Full dumps are required often and are too large.
The benefits of this reduction in scope are:
* No changes required in `GameMemory`
* Less code in `MiniDumper` (everything wrapped in `
#ifndef DISABLE_GAMEMEMORY`)
* Less complexity and chance for things to go wrong
* Easier to review
* Gives us insights in how often the crashdumps are actually useful and if it's worth investing in the `GameMemory`/extended type dumps before going to full CrashPad or similar solutions.

Does this seem like a reasonable path forward?

xezon reviewed (approved):
Looking good.
